liblava-demo/generic_triangle.cpp

```diff
@@ -0,0 +1,174 @@
+/**
+ * @file         liblava-demo/generic_triangle.cpp
+ * @brief        Generic triangle demo
+ * @authors      Lava Block OÜ and contributors
+ * @copyright    Copyright (c) 2018-present, MIT License
+ */
+
+#include <cstddef>
+#include <demo.hpp>
+
+using namespace lava;
+
+//-----------------------------------------------------------------------------
+int main(int argc, char* argv[]) {
+    app app("generic lava triangle", { argc, argv });
+
+    if (!app.setup())
+        return error::not_ready;
+
+    // Initialize a lava triangle.
+    generic_mesh<>::ptr lava_triangle;
+    lava_triangle = generic_create_mesh<>(app.device, mesh_type::triangle);
+    if (!lava_triangle)
+        return error::create_failed;
+    auto& lava_triangle_data = lava_triangle->get_data();
+    lava_triangle_data.vertices.at(0).color = v4(1.f, 0.f, 0.f, 1.f);
+    lava_triangle_data.vertices.at(1).color = v4(0.f, 1.f, 0.f, 1.f);
+    lava_triangle_data.vertices.at(2).color = v4(0.f, 0.f, 1.f, 1.f);
+    lava_triangle_data.scale<>(0.5f);
+    lava_triangle_data.move({ 0.5f, 0, 0 });
+    if (!lava_triangle->reload())
+        return error::create_failed;
+
+    // Initialize an int triangle.
+    typedef struct {
+        std::array<int, 3> position;
+        lava::v4 color;
+    } int_vertex;
+    generic_mesh<int_vertex>::ptr int_triangle;
+    int_triangle = generic_create_mesh<int_vertex, int>(app.device, mesh_type::triangle, offsetof(int_vertex, position));
+    if (!int_triangle)
+        return error::create_failed;
+    auto& int_triangle_data = int_triangle->get_data();
+    int_triangle_data.vertices.at(0).color = v4(1.f, 0.5f, 0.5f, 1.f);
+    int_triangle_data.vertices.at(1).color = v4(0.5f, 1.f, 0.5f, 1.f);
+    int_triangle_data.vertices.at(2).color = v4(0.5f, 0.5f, 1.f, 1.f);
+    int_triangle_data.scale<int>(2, offsetof(int_vertex, position));
```

Comment:
Is it not possible to omit the PosType `int` if you can keep this info in the generic mesh?

Reply:
Do you mean instead of:
```c++
template<typename T = lava::vertex>
struct generic_mesh_data {
```
The generic struct could be defined as:
```c++
template<typename T = lava::vertex, typename PosType = float>
struct generic_mesh_data {
```
I can do that. Although I'm not sure if that's overall more typing or less typing for the end user. Would calling `scale()` happen more often than declaring an instance of 

If you mean can we somehow infer the data type of T's position vector just from knowing type `T` and `offsetof(T, position)`, I don't _think_ that's possible in C++20 without doing something very complicated [like this](https://github.com/alexpolt/luple). Maybe linking a type reflection library like that would be worth considering in the future? It can automatically get a list the types that are a struct's fields. We could then have their offsetof compared to the inputted offsetof, and see which comparison is equal, in order to figure out the type of the position vector. Maybe C++23 will finally add `reflexpr` so that this becomes easy. :smile_cat:

Reply:
Actually I just had an idea
```c++
    void scale(int factor) {
        using PosType=decltype(vertices[0].position);
        for (T& vertex : vertices) {
            for (size_t i = 0; i < 3; i++) {
                (*((PosType*)(&vertex + struct_pos_offset)))[i] *= factor;
            }
        }
```
This compiles, but doesn't run quite right yet. I think if we _assume_ that some `position` field exists, this can be done.

Reply:
This code is working for me. What do you think?
```c++
private:
    using PosVecType=decltype(vertices[0].position);

public:
    void scale(auto factor) {
        for (T& vertex : vertices) {
            for (size_t i = 0; i < 3; i++) {
                (*reinterpret_cast<PosVecType*>(
                    &vertex + offsetof(T, position)))[i] *= factor;
            }
        }
    }
```

Reply:
Wait .. but then we could just say:
```c++
    void scale(auto factor) {
        for (T& vertex : vertices) {
            for (size_t i = 0; i < 3; i++) {
                vertex.position[i] *= factor;
            }
        }
    }
```
That's not a solution at all >,<

Reply:
🚀 Nice! And the same here:

```c++
    template<typename PosType = float>
    void move(std::array<PosType, 3> offset) {
        for (T& vertex : vertices) {
            for (auto i = 0u; i < 3; ++i) {
                vertex.position[i] += offset[i];
            }
        }
    }
```

Reply:
I have a new idea. The user can pass in `typeid(vertex.position).name()`, and that can be demangled in different ways for compiling using GNU, Clang, or MSVC using `#define`s from CMake that would be hidden from the user.

Reply:
Yeah .. I guess I'm not super thrilled that it forces a naming convention on users, but everything else I can think of is really weird and not user-friendly. For now, I'm good with assuming a `position` field exists, since it seems you are. Going that route,  I don't think templates are needed for `move()`, either.
```c++
    void move(std::array<std::any, 3> offset) {
        for (T& vertex : vertices) {
            for (size_t i = 0; i < 3; i++) {
                    vertex.position[i] += std::any_cast<PosType&>(offset[i]);
            }
        }
    }
```

Reply:
That could probably work too, but I find the template version easier to read

Reply:
Alright! I'll keep it templated then.